<HTML>
<HEAD>
<TITLE>TRANSIT documentation</TITLE>
<style>
.th {
line-height: 0;
display: inline-block;
border: solid 4px #FFFFFF;
width: 800px;
height: 500px;
box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.2);
transition: all 200ms ease-out;
}

.param{
padding-bottom: 15px;
}

.thumb{
width: 100%;
height: 100%;
}
</style>
</HEAD>

<BODY>


<H1>TRANSIT v1.2.32 Documentation</H1>

<br>


<H2>Table of Contents</H2>

<UL>
    <LI> <A HREF="#overview">Overview</A>
    <LI> <A HREF="#installation">Installation</A>
    <LI> <A HREF="#running">Running TRANSIT</A>
    <UL>
        <LI> <A HREF="#running_gui">GUI Mode</A>
        <LI> <A HREF="#running_console">Command line Mode</A>
    </UL>
    <LI> <A HREF="#analysis">Analysis Methods</A>
    <UL>
        <LI> <A HREF="#gumbel">Gumbel</A>
        <LI> <A HREF="#hmm">HMM</A>
        <LI> <A HREF="#resampling">Re-sampling</A>
    </UL>
    <LI> <A HREF="#tutorial_glyc_chol">Tutorial: Comparative Analysis - Glycerol vs Cholesterol</A>
    <UL>
        <LI> <A HREF="#glyc_chol_run">Run TRANSIT</A>
        <LI> <A HREF="#glyc_chol_annotation">Add annotation file</A>
        <LI> <A HREF="#glyc_chol_ctrl">Adding the control datasets</A>
        <LI> <A HREF="#glyc_chol_track">Visualizing read counts</A>
        <LI> <A HREF="#glyc_chol_scatter">Scatter plot</A>
        <LI> <A HREF="#glyc_chol_exp">Adding the experimental datasets</A>
        <LI> <A HREF="#glyc_chol_gumbel">Essentiality analysis using the Gumbel method</A>
        <LI> <A HREF="#glyc_chol_hmm">Essentiality analysis using the HMM method</A>
        <LI> <A HREF="#glyc_chol_resampling">Comparative analysis using Re-sampling</A>
        <LI> <A HREF="#glyc_chol_resampling_results">Viewing resampling results</A>
    </UL>
    <LI> <A HREF="#tutorial_export">Tutorial: Export Reads to IGV</A>
    <UL>
        <LI><A HREF="#export_add_annotation">Adding the annotation file</A>
        <LI> <A HREF="#export_add_datasets">Add .wig datasets</A>
        <LI> <A HREF="#export_to_IGV">Export to IGV</A>
    </UL>
</UL>




<P>
<BR>
<BR>
<BR>
<HR>


<H2><A id="overview">Overview</A></H2>

<UL>
<LI class="param"> This is a software that can be used to analyze Tn-Seq datasets.
It includes various statistical calculations of essentiality of genes or genomic regions (including conditional essentiality between 2 conditions). These methods were developed and tested as a collaboration between the Sassetti lab (UMass) and the Ioerger lab (Texas A&M).


<BR>
<BR>
<a class="th" href="images/tutorial_glyc_ctrl.png"><img src="images/tutorial_glyc_ctrl.png" class="thumb"/></a>
<BR>
<BR>





<LI class="param"> TRANSIT assumes you have already done pre-processing of raw sequencing files (.fastq) and extracted read counts (at each TA dinucleotide).  Actually, the <A HREF="http://www.springer.com/biomed/human+genetics/book/978-1-4939-2397-7">current protocol</A> utilizes internal barcodes that can be used to reduce raw read counts to unique template counts, and this this is the intended input to TRANSIT.  The input for TRANSIT consists of <A HREF="http://genome.ucsc.edu/goldenpath/help/wiggle.html">.wig</A> files, which simply list the coordinate of each TA sites in the genome and the number of templates observed.


<LI class="param"> 
There are various methods available for pre-processing (converting .fastq files to .wig files). You might have your own scripts (if so, massage the data into .wig format), or you might get the scripts used in the Sassetti lab.  For convenience, we are including a separate tool called <A HREF="http://saclab.tamu.edu/tom/TPP.html">TPP</A> (Tn-Seq Pre-Processor) with this distribution that encodes the way we process .fastq files in the Ioerger lab.  It's a complicated process with many steps (removing transposon prefixes of reads, mapping into genome, identifying barcodes and reducing read counts to template counts).

<LI class="param"> Most of the analysis methods in TRANSIT require an <B>annotation</B> to know the gene coordinates and names.  This is the top file input in the GUI window.  The annotation has to be in a somewhat non-standard format called a ".prot_table".  If you know what you are doing, it is easy to  convert annotations for other organsims into .prot_table format.  But for convenience, we are distributing the prot_tables for 3 common versions of the H37Rv genome: H37Rv.prot_table (NC_000962.2, from Stewart Cole), H37RvMA2.prot_table (sequenced version from the Sassetti lab), and H37RvBD.prot_table (sequenced by the Broad Institute).  All of these are slightly different, and it is <B>critical</B> that you use the same annotation file as the reference genome sequence used for mapping the reads (during pre-processing).

<LI class="param"> There are 2 main types of essentiality analyses: individaul, comaparative.  In individual analysis, the goal is to distinguish essential vs. non-essential in a single growth condition, and to assess the statistical significance of these calls.  Two methods for this are the Gumbel method and the HMM.  They are computationally distinct.  The Gumbel method is looking for significant stretches of TA sites lacking insertions, whereas the HMM looks for regions where the mean read count is locally suppresed or increased.  The HMM can detect 'growth-advantaged' and 'growth-defect' regions.  The HMM is also a bit more robust on low-density datasets (insertion density 20-30%).  But both methods have their merits and are complementary. For compararative analysis, TRANSIT uses 're-sampling', which is analogous to a permutation test, to determine if the sum of read counts differs significantly between two conditions.  Hence this can be used to identify conditionally essential regions and quantify the statistical significance.

<LI class="param"> TRANSIT has been designed to handle multiple replicates.  If you have two or more replicate dataset of the same libarary selected in the same condition, you can provide them, and more of the computational methods will do something reasonable with them.

<LI class="param"> For those methods that generate p-values, we often also calculate adjusted p-value (or 'q-values') which are corrected for multiple tests typically the Benjamini-Hochberg procedure.  A typical threshold for significance would be q&lt;0.05 (not p&lt;0.05).

<LI class="param"> It is important to understand the GUI model that TRANSIT uses It allows you to load up datasets (.wig files), select them, choose an analysis method, set parameters, and start the computation. It will generate <B>output files</B> in your local directory with the results.  These files can then be loaded into the interface and browser with custom displays and graphs.  The interface has 3 main windows or sections: 'Control Samples', 'Experimental Samples', 'Results Files.'  The first two are for loading input files ('Control Samples' would be like replicate datasets from a reference condition, like in vitro, rich media, etc.; 'Experimental Samples' would be where you would load replicates for a comparative conditions, like in vivo, or minimmal media, or low-iron, etc.)  The 'Results Files' section is initially empty, but after a computation finishes, it will automatically be populated with the corresponding output file. See the 'Tutorial' section below in this documentation for an illustraion of the overall process for a typical work-flow.

<LI class="param"> TRANSIT incorporates many interesting ways of looking at your data.

<UL>
<LI class="param"> Track view shows you a visual representation of the read counts at each site at a locus of interest (for selected datasets) somewhat like IGV.


<BR>
<BR>
<a class="th" href="images/track_view.png"><img src="images/track_view.png" class="thumb" /></a>
<BR>
<BR>


<LI class="param">Scatter plots can show the correlation of counts between 2 datasets.


<BR>
<BR>
<a class="th" href="images/scatter.png"><img src="images/scatter.png" class="thumb" /></a>
<BR>
<BR>



<LI class="param">Volcano plots can be used to visualize the results of resampling and assess the distribution between over- and under-represented genes in condition B vs. condition A. In addition you can look at histogram of the re-sample distributions for each gene.

<BR>
<BR>
<a class="th" href="images/result_volcano_graph.png"><img src="images/result_volcano_graph.png" class="thumb" /></a>
<BR>
<BR>


<BR>
<BR>
<a class="th" href="images/result_table_histogram.png"><img src="images/result_table_histogram.png" class="thumb" /></a>
<BR>
<BR>




</UL>

<LI class="param"> Most of the methods take a few minutes to run. (it depends on parameters, CPU clock speed, etc., but the point is, a) these calculations are complex and not instaneous, but b) we have tried to implement it so that they don't take hours)

<LI class="param"> Note: in this version of TRANSIT, most of the methods are oriented toward gene-level analysis.   There are methods for analyzing essentiality of arbitrary genomic regions (e.g. sliding windows, HMMs...).  We plan to incorporate some of these in future versions.

</UL>



<P>
<BR>
<BR>
<BR>
<HR>



<h2><A id="installation">Installation</A></H2>

TRANSIT is python-based.  Primarily, you have to have python installed (installed by default on most systems).  In addition, TRANSIT relies on some python packages/libraries/modules that you might need to install.  Fortunately, there is nothing to compile/make/build.  You can simply run it from the command line.

<H3>Requirements</H3>
The following libraries/modules are required to run TRANSIT:

<BR>
<UL>
    <li><a href="http://www.python.org/">Python 2.7</a>
    <li><a href="http://www.numpy.org/">Numpy 1.6.1</a>
    <li><a href="http://www.scipy.org/">Scipy 0.14.0</a>
    <li><a href="http://matplotlib.org/users/installing.html">matplotlib 1.1.1 (or higher)</a>
    <li><a href="http://www.wxpython.org/">wxpython 3.0.1</a> (for Mac OSX, use the <b>cocoa</b> version of wxPython)
    <li><a href="http://www.pythonware.com/products/pil/">PIL (Python Imaging Library)</a> or Pillow.
</UL>
<BR>

Install these using the appropriate methods for your machine, e.g. apt-get or yum (for unix machines) or pip or easy_install (for Macs) or installer file (for Windows).


<H2><A id="running">Running TRANSIT</A></H2>

<H3><A id="running_gui">GUI Mode</A></H3>
To run TRANSIT in GUI mode (should be the same on Linux, Windows and MacOS), from the command line run:

<PRE>
python PATH/src/transit.py
</PRE>

where PATH is the path to the TRANSIT installation directory. You might be able to double-click on icon for transit.py, if your OS associates .py files with python and automatically runs them.

Note, because TRANSIT has a graphical user interface, if you are trying to run TRANSIT across a network, for example, running on a unix server but displaying on a desktop machine, you will probably need to use 'ssh -Y' and a local X11 client (like Xming or Cygwin/X on PCs).




<H3><A id="running_console">Command line Mode</A></H3>

TRANSIT can also be run from the command line, without the GUI interface. This is convenient if you want to run many analyses in batch, as you can write a script that automatically runs that automatically runs TRANSIT from the command line.


<BR>
TRANSIT expects the user to specify which analysis method they wish to run. The user can choose from "gumbel", "hmm", or "resampling". By choosing a method, and adding the "-h" flag, you will get a list of all the necessary parameters and optional flags for the chosen method:
 
<PRE>python PATH/src/transit.py gumbel -h</PRE>



<BR>

<H4>Gumbel</H4>
<P>

To run the Gumbel analysis from the command line, type "python PATH/src/transit.py gumbel" followed by the following arguments:

<P>
<table border=1
<tr><th>Argument</th> <th>Type</th> <th>Description</th> <th>Default</th> <th>Example</th></tr>

<tr><td>annotation</td> <td>Required</td> <td>Path to annotation file in .prot_table format</td> <td></td> <td>genomes/H37Rv.prot_table</td></tr>
<tr><td>control_files</td> <td>Required</td> <td>Comma-separated list of paths to the *.wig replicate datasets</td> <td></td> <td>data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig</td></tr>

<tr><td>output_file</td> <td>Required</td> <td>Name of the output file with the results.</td> <td></td> <td>results/gumbel_glycerol.dat</td></tr>

<tr><td>-s SAMPLES</td> <td>Optional</td> <td>Number of samples to take.</td> <td>10000</td> <td>-s 20000</td></tr>
<tr><td>-m MINREAD</td> <td>Optional</td> <td>Smallest read-count considered to be an insertion.</td> <td>1</td> <td>-m 2</td></tr>
<tr><td>-b BURNIN</td> <td>Optional</td> <td>Burn in period, Skips this number of samples before getting estimates. See documentation.</td> <td>500</td> <td>-b 100</td></tr>
<tr><td>-t TRIM</td> <td>Optional</td> <td>Number of samples to trim. See documentation.</td> <td>1</td> <td>-t 2</td></tr>
<tr><td>-r REP</td> <td>Optional</td> <td>How to handle replicates read-counts: 'Sum' or 'Mean'.</td> <td>Sum</td> <td>-r Mean</td></tr>
<tr><td>-iN IGNOREN</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the N terminus.</td> <td>5</td> <td>-iN 0</td></tr>
<tr><td>-iC IGNOREC</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the C terminus.</td> <td>5</td> <td>-iC 10</td></tr>
</table>

<P>

Example Gumbel command:
<PRE>python PATH/src/transit.py gumbel genomes/H37Rv.prot_table data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig test_console_gumbel.dat -s 20000 -b 1000</PRE>



<BR>


<H4>HMM</H4>

To run the HMM analysis from the command line, type "python PATH/src/transit.py hmm" followed by the following arguments:

<P>
<table border=1
<tr><th>Argument</th> <th>Type</th> <th>Description</th> <th>Default</th> <th>Example</th></tr>

<tr><td>annotation</td> <td>Required</td> <td>Path to annotation file in .prot_table format</td> <td></td> <td>genomes/H37Rv.prot_table</td></tr>
<tr><td>control_files</td> <td>Required</td> <td>Comma-separated list of paths to the *.wig replicate datasets</td> <td></td> <td>data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig</td></tr>

<tr><td>output_file</td> <td>Required</td> <td>Name of the output file with the results.</td> <td></td> <td>results/gumbel_glycerol.dat</td></tr>

<tr><td>-r REP</td> <td>Optional</td> <td>How to handle replicates read-counts: 'Sum' or 'Mean'.</td> <td>Sum</td> <td>-r Mean</td></tr>
<tr><td>-iN IGNOREN</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the N terminus.</td> <td>5</td> <td>-iN 0</td></tr>
<tr><td>-iC IGNOREC</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the C terminus.</td> <td>5</td> <td>-iC 10</td></tr>
</table>



<P>

Example HMM command:
<PRE>python PATH/src/transit.py hmm genomes/H37Rv.prot_table data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig test_console_hmm.dat -r Sum</PRE>


<BR>



<H4>Resampling</H4>

To run the Resampling analysis from the command line, type "python PATH/src/transit.py resampling" followed by the following arguments:

<P>

<table border=1
<tr><th>Argument</th> <th>Type</th> <th width=60%>Description</th> <th>Default</th> <th>Example</th></tr>

<tr><td>annotation</td> <td>Required</td> <td>Path to annotation file in .prot_table format</td> <td></td> <td>genomes/H37Rv.prot_table</td></tr>

<tr><td>control_files</td> <td>Required</td> <td>Comma-separated list of paths to the *.wig replicate datasets for the control condition</td> <td></td> <td>data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig</td></tr>

<tr><td>exp_files</td> <td>Required</td> <td>Comma-separated list of paths to the *.wig replicate datasets for the experimental condition</td> <td></td> <td>data/cholesterol_reads_rep1.wig,data/cholesterol_reads_rep2.wig</td></tr>

<tr><td>output_file</td> <td>Required</td> <td>Name of the output file with the results.</td> <td></td> <td>results/gumbel_glycerol.dat</td></tr>

<tr><td>-s SAMPLES</td> <td>Optional</td> <td>Number of permutations performed.</td> <td>10000</td> <td>-s 5000</td></tr>

<tr><td>-H</td> <td>Optional</td> <td>Creates histograms of the permutations for all genes.</td> <td>Not set</td> <td>-H</td></tr>

<tr><td>-a</td> <td>Optional</td> <td>Performs adaptive appoximation to resampling.</td> <td>Not set</td> <td>-a</td></tr>

<tr><td>-dN</td> <td>Optional</td> <td>Uses raw counts, instead of normalizing between conditions.</td> <td>Not set</td> <td>-dN</td></tr>

<tr><td>-iN IGNOREN</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the N terminus.</td> <td>5</td> <td>-iN 0</td></tr>
<tr><td>-iC IGNOREC</td> <td>Optional</td> <td>Ignore TAs occuring at X% of the C terminus.</td> <td>5</td> <td>-iC 10</td></tr>
</table>


<P>
Example Resampling command:
<PRE>python PATH/src/transit.py resampling genomes/H37Rv.prot_table data/glycerol_reads_rep1.wig,data/glycerol_reads_rep2.wig data/cholesterol_reads_rep1.wig,data/cholesterol_reads_rep2.wig,data/cholesterol_reads_rep3.wig test_console_resampling.dat -H -s 10000 </PRE>






<P>
<BR>
<BR>
<BR>
<HR>


<H2><A id="analysis">Analysis Methods</A></H2>

<H3><A id="gumbel">Gumbel</A></H3>

<H4>What does it do?</H4>
The Gumbel can be used to determine which genes are essential in a single condition.  It does a gene-by-gene analysis of the insertions at TA sites with each gene, makes a call based on the longest consecutive sequence of TA sites without insertion in the genes, calculates the probabily of this using a Baysian model.

<H4>How does it work?</H4> 

For a formal description of how this method works, see our paper:
<BR>
DeJesus, M.A., Zhang, Y.J., Sassettti, C.M., Rubin, E.J., Sacchettini,
J.C., and Ioerger,
T.R. (2013). <BR><A HREF="http://www.ncbi.nlm.nih.gov/pubmed/23361328">Bayesian analysis of gene essentiality based on sequencing of transposon insertion libraries.</A> <I>Bioinformatics</I>, 29(6):695-703.


<H4>Parameters</H4>

<UL class="param">
    <LI class="param"><B>Samples:</B>  Gumbel uses Metropolis-Hastings (MH) to generate samples of posterior distributions. The default setting is to run the simulation for 10,000 iterations.  This is usually enough to assure convergence of the sampler and to provide accurate estimates of posterior probabilities. Less iterations may work, but at the risk of lower accuracy.
    
    <LI class="param"><B>Burn-In:</B>  Because the MH sampler many not have stabalized in the first few iterations, a "burn-in" period is defined. Samples obtained in this "burn-in" period are discarded, and do not count towards estimates.

    <LI class="param"><B>Trim:</B>  The MH sampler produces Markov samples that are correlated. This parameter dictates how many samples must be attempted for every sampled obtained. Increasing this parameter will decrease the auto-correlation, at the cost of dramatically increasing the run-time. For most situations, this parameter should be left at the default of "1".

    <LI class="param"><B>Minimum Read:</B> The minimum read count that is considered a true read. Because the Gumbel method depends on determining gaps of TA sites lacking insertions, it may be suceptible to spurious reads (e.g. errors). The default value of 1 will consider all reads as true reads. A value of 2, for example, will ignore read counts of 1.

    <LI class="param"><B>Replicates:</B> Determines how to deal with replicates by averaging the read-counts or suming read counts accross datasets. This should not have an affect for the Gumbel method, aside from potentially affecting spurious reads.
</UL>

<H4>Outputs and diagnostics</H4>

The Gumbel method generates a tab-seperated output file at the location chosen by the user. This file will automatically be loaded into the Results Files section of the GUI, allowing you to display it as a table. Alternatively, the file can be opened in a spreadsheet software like Excel as a tab-separated file. The columns of the output file are defined as follows:

<P>
<table border=1>
<TR><TH width="20%" align="left">Column Header</TH> <TH width="80%" align="left">Column Definition</TH></TR>
<TR><TD>ORF</TD> <TD>Gene ID.</TD> </TR>
<TR><TD>Name</TD> <TD>Name of the gene.</TD> </TR>
<TR><TD>Description</TD> <TD>Gene description.</TD> </TR>
<TR><TD>k</TD> <TD>Number of Transposon Insertions Observed within the ORF.</TD></TR>
<TR><TD>n</TD> <TD>Total Number of TA dinucleotides within the ORF.</TD></TR>
<TR><TD>r</TD> <TD>Length of the Maximum Run of Non-Insertions observed.</TD></TR>
<TR><TD>s</TD> <TD>Span of nucleotidies for the Maximum Run of Non-Insertions.</TD></TR>
<TR><TD>zbar</TD> <TD>Posterior Probability of Essentiality.</TD></TR>
<TR><TD>Call</TD> <TD>Essentiality call for the gene. Depends on FDR corrected thresholds. E=Essential U=Uncertain, NE=Non-Essential, S=too short</TD></TR>
</table>
<BR>
<BR>



Note: Technically, Bayesian models are used to calculate posterior probabilities, not p-values (which is a concept associated with the
frequentist framework).  However, we have implemented a method for computing the approximate false-discovery rate (FDR) that serves a 
similar purpose.  This determines a threshold for significance on the posterior probabilities that is corrected for multiple tests.  The actual thresholds used are reported in the headers of the output file (and are near 1 for essentials and near 0 for non-essentials). There can be many genes that score between the two thresholds (t1 &lt; zbar &lt; t2).  This reflects intrinsic uncertainty associated with either low read counts, sparse insertion density, or small genes.

If the insertion_density is too low (&lt; ~30%), the method may not work as well, and might indicate an unusually large number of Uncertain or Essential genes.


<H4>Run-time</H4>

The Gumbel method takes on the order of 10 minutes for 10,000 samples.  Run-time is linearly proportional to the 'samples' parameter, or length of MH sampling trajectory.

Other notes: Gumbel can be run on multiple replicates; replicate datasets will be automatically merged.


    

<P>
<BR>
<BR>
<BR>
<HR>



<H3><A id="hmm">HMM</A></H3>

<H4>What does it do?</H4>

The HMM method can be used to determine the essentiality of the entire genome, as opposed to gene-level analysis of the other methods. It is capable of identifying regions that have unusually high or unusually low read counts (i.e. growth advantage or growth defect regions), in addition to the more common categories of essential and non-essential.

<H4>How does it work?</H4>

For a formal description of how this method works, see our paper:
<BR>
DeJesus, M.A., Ioerger, T.R. <A HREf="http://www.ncbi.nlm.nih.gov/pubmed/24103077">A Hidden Markov Model for identifying essential and growth-defect regions in bacterial genomes from transposon insertion sequencing data.</A> <I>BMC Bioinformatics.</I> 2013. 14:303

<H4>Parameters</H4>

The HMM method automatically estimates the necessary statistical parameters from the datasets. You can change how the method handles replicate datasets:


<UL>
    <LI class="param"><B>Replicates:</B> Determines how the HMM deals with replicate datasets by either averaging the read-counts or suming read counts accross datasets. For regular datasets (i.e. mean-read count > 100) the recommended setting is to average read-counts together. For sparse datasets, it suming read-counts may produce more accurate results.
</UL>



<H4>Output and Diagnostics</H4> The HMM method outputs two files. The first file provides the most likely assignment of states for all the TA sites in the genome. Sites can belong to one of the following states: "E" (Essential), "GD" (Growth-Defect), "NE" (Non-Essential), or "GA" (Growth-Advantage). In addition, the output includes the probability of the particular site belonging to the given state. The columns of this file are defined as follows:

<BR>
<BR>
<table border=1>
<TR><TH width="20%" align="left">Column #</TH> <TH width="40%" align="left">Column Definition</TH></TR>
<TR><TD>1</TD> <TD>Coordinate of TA site</TD></TR>
<TR><TD>2</TD> <TD>Observed Read Counts</TD></TR>
<TR><TD>3</TD> <TD>Probability for ES state</TD></TR>
<TR><TD>4</TD> <TD>Probability for GD state</TD></TR>
<TR><TD>5</TD> <TD>Probability for NE state</TD></TR>
<TR><TD>6</TD> <TD>Probability for GA state</TD></TR>
<TR><TD>7</TD> <TD>State Classification (ES = Essential, GD = Growth Defect, NE = Non-Essential, GA = Growth-Defect)</TD></TR>
<TR><TD>8</TD> <TD>Gene(s) that share(s) the TA site.</TD></TR>
</table>
<BR>
<BR>


The second file provides a gene-level classification for all the genes in the genome. Genes are classified as "E" (Essential), "GD" (Growth-Defect), "NE" (Non-Essential), or "GA" (Growth-Advantage) depending on the number of sites within the gene that belong to those states.


<BR>
<BR>
<table border=1>
<TR><TH width="20%" align="left">Column Header</TH> <TH width="40%" align="left">Column Definition</TH></TR>
<TR><TD>Orf</TD> <TD>Gene ID</TD></TR>
<TR><TD>Name</TD> <TD>Gene Name</TD></TR>
<TR><TD>Desc</TD> <TD>Gene Description</TD></TR>
<TR><TD>N</TD> <TD>Number of TA sites</TD></TR>
<TR><TD>n0</TD> <TD>Number of sites labeled ES (Essential)</TD></TR>
<TR><TD>n1</TD> <TD>Number of sites labeled GD (Growth-Defect)</TD></TR>
<TR><TD>n2</TD> <TD>Number of sites labeled NE (Non-Essential)</TD></TR>
<TR><TD>n3</TD> <TD>Number of sites labeled GA (Growth-Advantage)</TD></TR>
<TR><TD>Avg. Insertions</TD> <TD>Mean insertion rate within the gene</TD></TR>
<TR><TD>Avg. Reads</TD> <TD>Mean read count within the gene</TD></TR>
<TR><TD>State Call</TD> <TD>State Classification (ES = Essential, GD = Growth Defect, NE = Non-Essential, GA = Growth-Defect)</TD></TR>
</table>
<BR>
<BR>


Note: Libraries that are too sparse (e.g. &lt; 30%) or which contain very low read-counts may be problematic for the HMM method, causing it to label too many Growth-Defect genes. 

 
<H4>Run-time</H4>

The HMM method takes less than 10 minutes to complete. The parameters of the method should not affect the running-time.



<BR>
<BR>
<BR>
<BR>
<HR>

<H3><A id="resampling">Re-sampling</A></H3>

<H4>What does it do?</H4>
The re-sampling method is a comparative analysis the allows that can be used to determine conditional essentiality of genes. It is based on a permutation test, and is capable of determining read-counts that are significantly different across conditions.

<H4>How does it work?</H4>
This technique has yet to be formally published in the context of differential essentiality analysis. Briefly, the read-counts at each genes are determined for each replicate of each condition. The total read-counts in condition A is substracted from the total read counts at condition B, to obtain an observed difference in read counts. The TA sites are then permuted for a given number of "samples". For each one of these permutations, the difference is read-counts is determined. This forms a null disttribution, from which a p-value is calculated for the original, observed difference in read-counts.


<H4>Parameters</H4>
The resampling method is non-parametric, and therefore does not require any parameters governing the distributions or the model. The following parameters are available for the method:

<UL>
    <LI class="param"><B>Samples:</B> The number of samples (permutations) to perform. The larger the number of samples, the more resolution the p-values calculated will have, at the expense of longer computation time. The re-sampling method runs on 10,000 samples by default.
    <LI class="param"><B>Output Histograms:</B>Determines whether to output .png images of the histograms obtained from resampling the difference in read-counts.
    <LI class="param"><B>Adaptive Resampling:</B> An optional "adaptive" version of resampling which accelerates the calculation by terminating early for genes which are likely not significant. This dramatically speeds up the computation at the cost of less accurate estimates for those genes  that terminate early (i.e. deemed not significant). This option is OFF by default.

</UL>




<H4>Output and Diagnostics</H4>
The re-sampling method ouputs a tab-delimited file with results for each gene in the genome. P-values are adjusted for multiple comparisons using the Benjamini-Hochberg procedure (called "q-values" or "p-adj."). A typical threshold for conditional essentiality on is q-value &lt; 0.05.



<P>
<table border=1>
<TR><TH width="20%" align="left">Column Header</TH> <TH width="80%" align="left">Column Definition</TH></TR>
<TR><TD>Orf</TD> <TD>Gene ID.</TD> </TR>
<TR><TD>Name</TD> <TD>Name of the gene.</TD> </TR>
<TR><TD>Description</TD> <TD>Gene description.</TD> </TR>
<TR><TD>N</TD> <TD>Number of TA sites in the gene.</TD></TR>
<TR><TD>TAs Hit</TD> <TD>Number of TA sites with at least one insertion.</TD></TR>
<TR><TD>Sum Rd 1</TD> <TD>Sum of read counts in condition 1.</TD></TR>
<TR><TD>Sum Rd 2</TD> <TD>Sum of read counts in condition 2.</TD></TR>
<TR><TD>Delta Rd</TD> <TD>Difference in the sum of read counts.</TD></TR>
<TR><TD>p-value</TD> <TD>P-value calculated by the permutation test.</TD></TR>
<TR><TD>p-adj.</TD> <TD>Adjusted p-value controlling for the FDR (Benjamini-Hochberg)</TD></TR>
</table>
<BR>
<BR>





<H4>Run-time</H4>
A typical run of the re-sampling method with 10,000 samples will take around 45 minutes (with the histogram option ON). Using the adaptive resampling option, the run-time is reduced to around 10 minutes.







<P>
<BR>
<BR>
<BR>
<HR>



<H2><A id="tutorial_glyc_chol">Tutorial: Essentiality Analysis - Glycerol and Cholesterol</A></H2>
To illustrate how TRANSIT works, we are going to go through a tutorial where we analyze datasets of H37Rv <I>M. tuberculosis</I> grown on glycerol and cholesterol. 

<H3><A id="glyc_chol_run">Run TRANSIT</A></H3>
Navigate to the directory containing the TRANSIT files, and run TRANSIT:


<PRE>
python PATH/src/transit.py
</PRE>



<H3><A id="glyc_chol_annotation">Adding the annotation file</A></H3>
Before we can analyze datasets, we need to add an annotation file for the organism corresponding to the desired datasets. Click on the file dialog button, on the top of the TRANSIT window (see image below), and browse and select the appropriate annotation file. Note: Annotation files must be in ".prot_table" format, described above.

<BR>
<BR>
<a class="th" href="images/tutorial_annotation.png"><img class="thumb" src="images/tutorial_annotation.png"/></a>
<BR>
<BR>


<H3><A id="glyc_chol_ctrl">Adding the control datasets</A></H3>

We want to analyze datasets grown in glycerol to those grown in cholesterol. We are choosing the datasets grown in glycerol as the "Control" datasets. To add these, we click on the control sample file dialog (see image below), and select the desired datasets (one by one). In this example, we have two replicates:

<BR>
<BR>
<a class="th" href="images/tutorial_glyc_ctrl.png"><img src="images/tutorial_glyc_ctrl.png" class="thumb"/></a>
<BR>
<BR>

As we add the datasets they will appear in the table in the Control samples section. This table will provide the following statistics about the datasets that have been loaded so far: Total Number of Reads, Density, Mean Read Count and Maximum Count. These statistics can be used as general diagnostics of the datasets.


<H3><A id="glyc_chol_track">Visualizing read counts</A></H3>
TRANSIT allows us to visualize the read-counts of the datasets we have already loaded. To do this, we must select the desired datasets ("Control+Click") and then click on "View -> Track View" in the menu bar at the top of the TRANSIT window. Only those selected datasets will be displayed:

<BR>
<BR>
<a class="th" href="images/tutorial_view_menu.png"><img src="images/tutorial_view_menu.png" class="thumb"/></a>
<BR>
<BR>

This will open a window that allows that shows a visual representation of the read counts at the TA sites throughout the genome. The scale of the read counts can be set by changing the value of the "Max Read" textbox on the right. We can browse around the genome by clicking on the left and right arrowm, or search for a specific gene with the search text box.

<BR>
<BR>
<a class="th" href="images/tutorial_track_view.png"><img src="images/tutorial_track_view.png" class="thumb" /></a>
<BR>
<BR>


This window also allows us to save a .png image of the canvas for future reference if desired (i.e. Save Img button).





<H3><A id="glyc_chol_scatter">Scatter plot</A></H3>
We can also view a scatter plot of read counts of two selected datasets. To achieve this we select two datasets (using "Control + Clicck") and then clicking on "View -> Scatter Plot" in the menu bar at the top of the TRANSIT window.


<BR>
<BR>
<a class="th" href="images/tutorial_view_menu.png"><img src="images/tutorial_view_menu.png" class="thumb" /></a>
<BR>
<BR>


A new window will pop-up, show a scatter plot of both of the selected datasets. This window contains controls to zoom in and out (magnifying glass), allowing us to focus in on a specific area. This is particularly useful when large outliers may throw off the scale of the scatter plot.


<BR>
<BR>
<a class="th" href="images/tutorial_scatter.png"><img src="images/tutorial_scatter.png" class="thumb" /></a>
<BR>
<BR>




<H3><A id="glyc_gumbel">Essentiality analysis with the Gumbel method</A></H3>
Before comparing both conditions against each other, we may want to determine which genes are essential in a specific condition to get an idea of the genes which are required. To do this we can use the Gumbel or the HMM methods, which determine essentiality within one condition.

First we chose the Gumbel method from the drop-down menu on the right hand side of the TRANSIT window:

<BR>
<BR>
<a class="th" href="images/tutorial_gumbel_menu.png"><img src="images/tutorial_gumbel_menu.png" class="thumb"/></a>
<BR>
<BR>


For this particular case we leave the parameters at their default settings as these work with a wide variety of datasets (See above for an explanation of their function). We then click on the "Run Gumbel" button and wait until the analysis finished running. The progress bar will give us information about how much of the analysis is still left.


Once the program finishes, the results file is automatically created (with the name chosen at run-time) and it is automatically added to the Results File section at the bottom of TRANSIT. We can visualize the results by selecting this file from the list, and clicking on the "Display Table" button. This will open a new window with a table of resuls:



<BR>
<BR>
<a class="th" href="images/tutorial_gumbel_file.png"><img src="images/tutorial_gumbel_file.png" class="thumb" /></a>
<BR>
<BR>



From this window we can view results, and sort on a specific column (described above) by clicking on a column header. In addition, the top of this window contains a breakdown of the number of essential and non-essential genes found by the Gumbel method. We can see that 675 genes are found to be essential by the Gumbel method (16%), roughly matching expectations that 15% of the genomes is necessary for growth in bacterial organisms. Clicking on the "Zbar" column we can sort the data on the posterior probability of essentiality. If we sort in descending order, we get those genes which are most likely to be essential on the top. Among these are genes like GyrA (DNA gyrase A) and RpoB (DNA-directed polymerase), which are both well-known essential genes, and which are obtain a posterior probability of essentiality of 1.0 (Essential).





<H3><A id="glyc_hmm">Essentiality analysis with the HMM  method</A></H3>

An alternative method for determining essentiality is the HMM method. This method differs from the Gumbel method in that is capable of assessing the essentiality of the entire genome, and is not limited to a gene-level analysis (See above for discussions of the pros and cons of each method). To run the HMM method we select it from the drop-down menu to the right. This automatically displays the available options for the HMM methid. Because the HMM method estimates parameters by examining the datasets, there is no need to set parameters for the model. 

One important option provided is how to deal with replicate datasets. Because the glycerol replicates had a mean read-count between 53-85, we decide to sum read-counts together by selecting "Sum" from the drop-down option.



<BR>
<BR>
<a class="th" href="images/tutorial_hmm_menu.png"><img src="images/tutorial_hmm_menu.png" class="thumb"/></a>
<BR>
<BR>


Finally we click on the "Run HMM" button, and wait for the method to finish. Once the analysis finishes, two new files will be created and automatically added to the list of files in the Results Files section. One file contains the output of states for each TA site in the genome. The other file contains the analysis for each gene. We can display each of the files be selecting them (individually) and clicking on the "Display Table" button (one at a time).

<P>
Like for the Gumbel method, a break down of the states is provided at the top of the table. In the case of glycerol, the HMM analysis classifies 16.3% of the genome as belonging to the "Essential" state, 5.4% belonging to the Growth-Defect state, 77.1% to the Non-Essential state, and 1.2% to the Growth Advantage state. This break down can be used as a diagnostic, to see if the results match our expectations. For example, in datasets with very low read-counts, or very low density, the percentage of Growth-Defect states may be higher (e.g. &gt; 10% ), which could indicate a problem.


<BR>
<BR>
<a class="th" href="images/tutorial_hmm_sites.png"><img src="images/tutorial_hmm_sites.png" class="thumb"/></a>
<BR>
<BR>




The HMM sites file contains the state assignments for all the TA sites in the genome. This file is particularly useful to browse for browsing the different types of regions in the genome. We can use this file to see how regions have different impacts on the growth-advantage (or disadvantage) of the organism. For example, the PDIM locus, which is required for virulance in vivo, results in a Growth-Advantage for the organism when disrupted. We can see this in the HMM Sites file by scrolling down to this region (Rv2930-Rv2939) and noticing the large read-counts at these sites, and the how they are labeled "GA".







<H3><A id="glyc_chol_exp">Adding the experimental datasets</A></H3>
We now repeat the process we did for control samples, for the experimental datasets that were grown on cholesterol. To add these, we click on the experimental sample file dialog (see image below), and select the desired datasets (one by one). In this example, we have three replicates:

<BR>
<BR>
<a class="th" href="images/tutorial_chol_exp.png"><img src="images/tutorial_chol_exp.png" class="thumb" /></a>
<BR>
<BR>




<H3><A id="glyc_chol_resampling">Comparative analysis using Re-sampling</A></H3>
To compare the growth conditions and assess conditional essentiality, we select "Resampling" from the list of methods in the drop-down menu on the right side of the TRANSIT window:


<BR>
<BR>
<a class="th" href="images/tutorial_resampling.png"><img src="images/tutorial_resampling.png" class="thumb" /></a>
<BR>
<BR>


This will populate the right side with options specific to the Resampling method. In this case, we choose to proceed with the default settings. However, we could have set a different number of samples for the resampling method or chosen the "Adaptive Resampling" option if we were interested in quicker results. See the description of the method above for more information.

<P>

<BR>
<BR>
<a class="th" href="images/tutorial_resampling_running.png"><img src="images/tutorial_resampling_running.png" class="thumb" /></a>
<BR>
<BR>


We click on the "Run Resampling" button to start the analysis. This will take several minutes to finish. The progress bar will give us an idea of how much time is left.



<H3><A id="glyc_chol_resampling_results">Viewing resampling results</A></H3>
Once TRANSIT finishes running, the results file will automatically be added to the Results Files section at the bottom of the window:

<BR>
<BR>
<a class="th" href="images/tutorial_resampling_result.png"><img src="images/tutorial_resampling_result.png" class="thumb" /></a>
<BR>
<BR>


This window allows us to track the results files that have been created in this session. From here, we can display a volcano plot of the resampling results by selecting the file from the list and clicking on the "Display Graph" button. This will open a new window containing the figure:


<BR>
<BR>
<a class="th" href="images/result_volcano_graph.png"><img src="images/result_volcano_graph.png" class="thumb" /></a>
<BR>
<BR>



To view the actual results, we can open the file in a new window by selecting it from the list and clicking on the "Display Table" button:


<BR>
<BR>
<a class="th" href="images/tutorial_resampling_file.png"><img src="images/tutorial_resampling_file.png" class="thumb" /></a>
<BR>
<BR>


 The newly opened window will display a table of the results. We can sort the results by clicking on the column header. For example, to focus on the genes that are most likely to be conditionally essential between glycerol and cholesterol, we can click on the column header labeled "q-value", which represents p-values that have been adjusted for multiple comparisons.

<P>

 Sorting q-values in ascending order, we can see those genes which are most likely to be conditionally essential on the top. A typical threshold for significance is &lt; 0.05.  We can use "Delta Sum" column to see which conditions had the most read counts in a particular gene. The sign of this value (+/-) lets us know on which condition the gene is essential and which condition it is non-essential. The magnitude lets us know how large the difference is. For example, glycerol kinase (GlpK) is necessary for growth on glycerol but it is not expected to be necessary when grown on another carbon source like cholesterol. We confirm our expectations by noticing that the sum of read counts in glycerol is only 22 reads (normalized), while there are a total of 2119 reads in cholesterol. The difference (2096) is positive, which means it is necessary for growth in glycerol but not cholesterol. 


<P>

Because we ran the resampling method with the "Histograms" options, we also have the ability to view the histograms of permutation differences for each of the genes by selecting a gene and right clicking:


<BR>
<BR>
<a class="th" href="images/result_table_menu.png"><img src="images/result_table_menu.png" class="thumb" /></a>
<BR>
<BR>

From this menu we can display the histogram, or view the read-counts for that specific gene in Track View:

<BR>
<BR>
<a class="th" href="images/result_table_histogram.png"><img src="images/result_table_histogram.png" class="thumb" /></a>
<BR>
<BR>







<P>
<BR>
<BR>
<BR>
<HR>





<H2><A id="tutorial_export">Tutorial: Export Reads to IGV</A></H2>




<H3><A id="export_add_annotation">Adding the annotation file</A></H3>
Before we can export .wig datasets to IGV format, we need to add an annotation file for the organism. Click on the file dialog button, on the top of the TRANSIT window (see image below), and browse and select the appropriate annotation file. Note: Annotation files must be in ".prot_table" format, described above:

<BR>
<BR>
<a class="th" href="images/tutorial_annotation.png"><img src="images/tutorial_annotation.png" class="thumb" /></a>
<BR>
<BR>



<H3><A id="export_add_datasets">Add .wig datasets</A></H3>

Next we must choose to add .wig formatted datasets what we wish to export to IGV format. To add these, we click on the control sample file dialog (see image below), and select the desired datasets (one by one). In this example, we have two replicates:

<BR>
<BR>
<a class="th" href="images/tutorial_glyc_ctrl.png"><img src="images/tutorial_glyc_ctrl.png" class="thumb" /></a>
<BR>
<BR>

As we add the datasets they will appear in the table below.


<H3><A id="export_to_IGV">Export to IGV</A></H3>

Finally, to export the datasets we click on "Export" in the menu bar at the top of the TRANSIT window, and select the option that matches which datasets we wish to export. Note: Only selected datasets ("Control+Click") will be exported.

<BR>
<BR>
<a class="th" href="images/export_menu.png"><img src="images/export_menu.png" class="thumb" /></a>
<BR>
<BR>



