<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial: Essentiality Analysis of the Entire Genome &mdash; transit v2.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'v2.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="transit v2.0.0 documentation" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial-essentiality-analysis-of-the-entire-genome">
<h1>Tutorial: Essentiality Analysis of the Entire Genome<a class="headerlink" href="#tutorial-essentiality-analysis-of-the-entire-genome" title="Permalink to this headline">¶</a></h1>
<p>To illustrate how TRANSIT works, we are going to go through a tutorial
where we analyze datasets of H37Rv M. tuberculosis grown on glycerol
and cholesterol.</p>
<div class="section" id="run-transit">
<h2>Run TRANSIT<a class="headerlink" href="#run-transit" title="Permalink to this headline">¶</a></h2>
<p>Navigate to the directory containing the TRANSIT files, and run
TRANSIT:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">PATH</span><span class="o">/</span><span class="n">src</span><span class="o">/</span><span class="n">transit</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-the-annotation-file">
<h2>Adding the annotation file<a class="headerlink" href="#adding-the-annotation-file" title="Permalink to this headline">¶</a></h2>
<p>Before we can analyze datasets, we need to add an annotation file for
the organism corresponding to the desired datasets. Click on the file
dialog button, on the top of the TRANSIT window (see image below), and
browse and select the appropriate annotation file. Note: Annotation
files must be in &#8221;.prot_table&#8221; format, described above.</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_annotation.png"><img alt="_images/transit_tutorial_annotation.png" class="align-center" src="_images/transit_tutorial_annotation.png" style="width: 600px;" /></a>
</div>
<div class="section" id="adding-the-control-datasets">
<h2>Adding the control datasets<a class="headerlink" href="#adding-the-control-datasets" title="Permalink to this headline">¶</a></h2>
<p>We want to analyze datasets grown in glycerol to those grown in
cholesterol. We are choosing the datasets grown in glycerol as the
&#8220;Control&#8221; datasets. To add these, we click on the control sample file
dialog (see image below), and select the desired datasets (one by
one). In this example, we have two replicates:</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_control_datasets.png"><img alt="_images/transit_tutorial_control_datasets.png" class="align-center" src="_images/transit_tutorial_control_datasets.png" style="width: 600px;" /></a>
<p>As we add the datasets they will appear in the table in the Control
samples section. This table will provide the following statistics
about the datasets that have been loaded so far: Total Number of
Reads, Density, Mean Read Count and Maximum Count. These statistics
can be used as general diagnostics of the datasets.</p>
</div>
<div class="section" id="visualizing-read-counts">
<h2>Visualizing read counts<a class="headerlink" href="#visualizing-read-counts" title="Permalink to this headline">¶</a></h2>
<p>TRANSIT allows us to visualize the read-counts of the datasets we have
already loaded. To do this, we must select the desired datasets
(&#8220;Control+Click&#8221;) and then click on &#8220;View -&gt; Track View&#8221; in the menu
bar at the top of the TRANSIT window. Only those selected datasets
will be displayed:</p>
<a class="reference internal image-reference" href="_images/transit_dataset_track_view.png"><img alt="_images/transit_dataset_track_view.png" class="align-center" src="_images/transit_dataset_track_view.png" style="width: 600px;" /></a>
<p>This will open a window that allows that shows a visual representation
of the read counts at the TA sites throughout the genome. The scale of
the read counts can be set by changing the value of the &#8220;Max Read&#8221;
textbox on the right. We can browse around the genome by clicking on
the left and right arrowm, or search for a specific gene with the
search text box.</p>
<p>This window also allows us to save a .png image of the canvas for
future reference if desired (i.e. Save Img button).</p>
</div>
<div class="section" id="scatter-plot">
<h2>Scatter plot<a class="headerlink" href="#scatter-plot" title="Permalink to this headline">¶</a></h2>
<p>We can also view a scatter plot of read counts of two selected
datasets. To achieve this we select two datasets (using &#8220;Control +
Clicck&#8221;) and then clicking on &#8220;View -&gt; Scatter Plot&#8221; in the menu bar
at the top of the TRANSIT window.</p>
<a class="reference internal image-reference" href="_images/transit_dataset_scatter_graph.png"><img alt="_images/transit_dataset_scatter_graph.png" class="align-center" src="_images/transit_dataset_scatter_graph.png" style="width: 600px;" /></a>
<p>A new window will pop-up, show a scatter plot of both of the selected
datasets. This window contains controls to zoom in and out (magnifying
glass), allowing us to focus in on a specific area. This is
particularly useful when large outliers may throw off the scale of the
scatter plot.</p>
</div>
<div class="section" id="essentiality-analysis-with-the-hmm-method">
<h2>Essentiality analysis with the HMM method<a class="headerlink" href="#essentiality-analysis-with-the-hmm-method" title="Permalink to this headline">¶</a></h2>
<p>An alternative method for determining essentiality is the HMM method.
This method differs from the Gumbel method in that is capable of
assessing the essentiality of the entire genome, and is not limited to
a gene-level analysis (See above for discussions of the pros and cons
of each method). To run the HMM method we select it from the list of
(Himar1) methods on the Analysis at the top. This automatically displays
the available options for the HMM methid. Because the HMM method estimates
parameters by examining the datasets, there is no need to set parameters
for the model. One important option provided is how to deal with replicate
datasets. Because the glycerol replicates had a mean read-count
between 53-85, we decide to sum read-counts together by selecting
&#8220;Sum&#8221; from the drop-down option.</p>
<a class="reference internal image-reference" href="_images/transit_hmm_options.png"><img alt="_images/transit_hmm_options.png" class="align-center" src="_images/transit_hmm_options.png" style="width: 600px;" /></a>
<p>Finally we click on the &#8220;Run HMM&#8221; button, and wait for the method to
finish. Once the analysis finishes, two new files will be created and
automatically added to the list of files in the Results Files section.
One file contains the output of states for each TA site in the genome.
The other file contains the analysis for each gene. We can display
each of the files be selecting them (individually) and clicking on the
&#8220;Display Table&#8221; button (one at a time).
Like for the Gumbel method, a break down of the states is provided at
the top of the table. In the case of glycerol, the HMM analysis
classifies 16.3% of the genome as belonging to the &#8220;Essential&#8221; state,
5.4% belonging to the Growth-Defect state, 77.1% to the Non-Essential
state, and 1.2% to the Growth Advantage state. This break down can be
used as a diagnostic, to see if the results match our expectations.
For example, in datasets with very low read-counts, or very low
density, the percentage of Growth-Defect states may be higher (e.g. &gt;
10% ), which could indicate a problem.</p>
<p>The HMM sites file contains the state assignments for all the TA sites
in the genome. This file is particularly useful to browse for browsing
the different types of regions in the genome. We can use this file to
see how regions have different impacts on the growth-advantage (or
disadvantage) of the organism. For example, the PDIM locus, which is
required for virulance in vivo, results in a Growth-Advantage for the
organism when disrupted. We can see this in the HMM Sites file by
scrolling down to this region (Rv2930-Rv2939) and noticing the large
read-counts at these sites, and the how they are labeled &#8220;GA&#8221;.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/transit_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial: Essentiality Analysis of the Entire Genome</a><ul>
<li><a class="reference internal" href="#run-transit">Run TRANSIT</a></li>
<li><a class="reference internal" href="#adding-the-annotation-file">Adding the annotation file</a></li>
<li><a class="reference internal" href="#adding-the-control-datasets">Adding the control datasets</a></li>
<li><a class="reference internal" href="#visualizing-read-counts">Visualizing read counts</a></li>
<li><a class="reference internal" href="#scatter-plot">Scatter plot</a></li>
<li><a class="reference internal" href="#essentiality-analysis-with-the-hmm-method">Essentiality analysis with the HMM method</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/transit_genome_tutorial.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Michael A. DeJesus.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>