<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Overview &mdash; transit v2.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'v2.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="transit v2.0.0 documentation" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">Â¶</a></h1>
<ul class="simple">
<li>This is a software that can be used to analyze Tn-Seq datasets. It includes various statistical calculations of essentiality of genes or genomic regions (including conditional essentiality between 2 conditions). These methods were developed and tested as a collaboration between the Sassetti lab (UMass) and the Ioerger lab (Texas A).</li>
</ul>
<a class="reference internal image-reference" href="_images/transit_interface.png"><img alt="_images/transit_interface.png" class="align-center" src="_images/transit_interface.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>TRANSIT assumes you have already done pre-processing of raw sequencing files (.fastq) and extracted read counts (at each TA dinucleotide). Actually, the <a class="reference external" href="http://www.springer.com/biomed/human+genetics/book/978-1-4939-2397-7">current protocol</a> utilizes internal barcodes that can be used to reduce raw read counts to unique template counts, and this this is the intended input to TRANSIT. The input for TRANSIT consists of <a class="reference external" href="http://genome.ucsc.edu/goldenpath/help/wiggle.html">.wig</a> files, which simply list the coordinate of each TA sites in the genome and the number of templates observed.</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>There are various methods available for pre-processing (converting .fastq files to .wig files). You might have your own scripts (if so, massage the data into .wig format), or you might get the scripts used in the Sassetti lab. For convenience, we are including a separate tool called <a class="reference external" href="http://saclab.tamu.edu/tom/TPP.html">TPP</a> (Tn-Seq Pre-Processor) with this distribution that encodes the way we process .fastq files in the Ioerger lab. It&#8217;s a complicated process with many steps (removing transposon prefixes of reads, mapping into genome, identifying barcodes and reducing read counts to template counts).</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>Most of the analysis methods in TRANSIT require an <strong>annotation</strong> to know the gene coordinates and names. This is the top file input in the GUI window. The annotation has to be in a somewhat non-standard format called a &#8221;.prot_table&#8221;. If you know what you are doing, it is easy to convert annotations for other organsims into .prot_table format. But for convenience, we are distributing the prot_tables for 3 common versions of the H37Rv genome: H37Rv.prot_table (NC_000962.2, from Stewart Cole), H37RvMA2.prot_table (sequenced version from the Sassetti lab), and H37RvBD.prot_table (sequenced by the Broad Institute). All of these are slightly different, and it is <strong>critical</strong> that you use the same annotation file as the reference genome sequence used for mapping the reads (during pre-processing).</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>There are 2 main types of essentiality analyses: individaul, comaparative. In individual analysis, the goal is to distinguish essential vs. non-essential in a single growth condition, and to assess the statistical significance of these calls. Two methods for this are the Gumbel method and the HMM. They are computationally distinct. The Gumbel method is looking for significant stretches of TA sites lacking insertions, whereas the HMM looks for regions where the mean read count is locally suppresed or increased. The HMM can detect &#8216;growth-advantaged&#8217; and &#8216;growth-defect&#8217; regions. The HMM is also a bit more robust on low-density datasets (insertion density 20-30%). But both methods have their merits and are complementary. For compararative analysis, TRANSIT uses &#8216;re-sampling&#8217;, which is analogous to a permutation test, to determine if the sum of read counts differs significantly between two conditions. Hence this can be used to identify conditionally essential regions and quantify the statistical significance.</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>TRANSIT has been designed to handle multiple replicates. If you have two or more replicate dataset of the same libarary selected in the same condition, you can provide them, and more of the computational methods will do something reasonable with them.</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>For those methods that generate p-values, we often also calculate adjusted p-value (or &#8216;q-values&#8217;) which are corrected for multiple tests typically the Benjamini-Hochberg procedure. A typical threshold for significance would be q&lt;0.05 (not p&lt;0.05).</li>
<li>It is important to understand the GUI model that TRANSIT uses It allows you to load up datasets (.wig files), select them, choose an analysis method, set parameters, and start the computation. It will generate <strong>output files</strong> in your local directory with the results. These files can then be loaded into the interface and browser with custom displays and graphs. The interface has 3 main windows or sections: &#8216;Control Samples&#8217;, &#8216;Experimental Samples&#8217;, &#8216;Results Files.&#8217; The first two are for loading input files (&#8216;Control Samples&#8217; would be like replicate datasets from a reference condition, like in vitro, rich media, etc.; &#8216;Experimental Samples&#8217; would be where you would load replicates for a comparative conditions, like in vivo, or minimmal media, or low-iron, etc.) The &#8216;Results Files&#8217; section is initially empty, but after a computation finishes, it will automatically be populated with the corresponding output file. See the &#8216;Tutorial&#8217; section below in this documentation for an illustraion of the overall process for a typical work-flow.</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li>TRANSIT incorporates many interesting ways of looking at your data.</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li>Track view shows you a visual representation of the read counts at each site at a locus of interest (for selected datasets) somewhat like IGV.</li>
</ul>
</div></blockquote>
<a class="reference internal image-reference" href="_images/transit_dataset_track_view.png"><img alt="_images/transit_dataset_track_view.png" class="align-center" src="_images/transit_dataset_track_view.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li>Scatter plots can show the correlation of counts between 2 datasets.</li>
</ul>
</div></blockquote>
<a class="reference internal image-reference" href="_images/transit_dataset_scatter_graph.png"><img alt="_images/transit_dataset_scatter_graph.png" class="align-center" src="_images/transit_dataset_scatter_graph.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line">+ Volcano plots can be used to visualize the results of resampling and assess the distribution between over- and under-represented genes in condition B vs. condition A. In addition you can look at histogram of the re-sample distributions for each gene.</div>
</div>
<a class="reference internal image-reference" href="_images/transit_result_volcano_graph.png"><img alt="_images/transit_result_volcano_graph.png" class="align-center" src="_images/transit_result_volcano_graph.png" style="width: 600px;" /></a>
<a class="reference internal image-reference" href="_images/transit_resampling_histogram_graph.png"><img alt="_images/transit_resampling_histogram_graph.png" class="align-center" src="_images/transit_resampling_histogram_graph.png" style="width: 600px;" /></a>
<ul class="simple">
<li>Most of the methods take a few minutes to run. (it depends on parameters, CPU clock speed, etc., but the point is, a) these calculations are complex and not instaneous, but b) we have tried to implement it so that they don&#8217;t take hours)</li>
<li>Note: in this version of TRANSIT, most of the methods are oriented toward gene-level analysis. There are methods for analyzing essentiality of arbitrary genomic regions (e.g. sliding windows, HMMs...). We plan to incorporate some of these in future versions.</li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/transit_logo.png" alt="Logo"/>
            </a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/transit_overview.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Michael A. DeJesus.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>