<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial: Essentiality Analysis in a Single Condition &mdash; transit v2.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'v2.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="transit v2.0.0 documentation" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial-essentiality-analysis-in-a-single-condition">
<h1>Tutorial: Essentiality Analysis in a Single Condition<a class="headerlink" href="#tutorial-essentiality-analysis-in-a-single-condition" title="Permalink to this headline">¶</a></h1>
<p>To illustrate how TRANSIT works, we are going to go through a tutorial
where we analyze datasets of H37Rv M. tuberculosis grown on glycerol
and cholesterol.</p>
<div class="section" id="run-transit">
<h2>Run TRANSIT<a class="headerlink" href="#run-transit" title="Permalink to this headline">¶</a></h2>
<p>Navigate to the directory containing the TRANSIT files, and run
TRANSIT:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">PATH</span><span class="o">/</span><span class="n">src</span><span class="o">/</span><span class="n">transit</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-the-annotation-file">
<h2>Adding the annotation file<a class="headerlink" href="#adding-the-annotation-file" title="Permalink to this headline">¶</a></h2>
<p>Before we can analyze datasets, we need to add an annotation file for
the organism corresponding to the desired datasets. Click on the file
dialog button, on the top of the TRANSIT window (see image below), and
browse and select the appropriate annotation file. Note: Annotation
files must be in &#8221;.prot_table&#8221; format, described above.</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_annotation.png"><img alt="_images/transit_tutorial_annotation.png" class="align-center" src="_images/transit_tutorial_annotation.png" style="width: 600px;" /></a>
</div>
<div class="section" id="adding-the-control-datasets">
<h2>Adding the control datasets<a class="headerlink" href="#adding-the-control-datasets" title="Permalink to this headline">¶</a></h2>
<p>We want to analyze datasets grown in glycerol to those grown in
cholesterol. We are choosing the datasets grown in glycerol as the
&#8220;Control&#8221; datasets. To add these, we click on the control sample file
dialog (see image below), and select the desired datasets (one by
one). In this example, we have two replicates:</p>
<a class="reference internal image-reference" href="_images/transit_tutorial_control_datasets.png"><img alt="_images/transit_tutorial_control_datasets.png" class="align-center" src="_images/transit_tutorial_control_datasets.png" style="width: 600px;" /></a>
<p>As we add the datasets they will appear in the table in the Control
samples section. This table will provide the following statistics
about the datasets that have been loaded so far: Total Number of
Reads, Density, Mean Read Count and Maximum Count. These statistics
can be used as general diagnostics of the datasets.</p>
</div>
<div class="section" id="visualizing-read-counts">
<h2>Visualizing read counts<a class="headerlink" href="#visualizing-read-counts" title="Permalink to this headline">¶</a></h2>
<p>TRANSIT allows us to visualize the read-counts of the datasets we have
already loaded. To do this, we must select the desired datasets
(&#8220;Control+Click&#8221;) and then click on &#8220;View -&gt; Track View&#8221; in the menu
bar at the top of the TRANSIT window. Only those selected datasets
will be displayed:</p>
<a class="reference internal image-reference" href="_images/transit_dataset_track_view.png"><img alt="_images/transit_dataset_track_view.png" class="align-center" src="_images/transit_dataset_track_view.png" style="width: 600px;" /></a>
<p>This will open a window that allows that shows a visual representation
of the read counts at the TA sites throughout the genome. The scale of
the read counts can be set by changing the value of the &#8220;Max Read&#8221;
textbox on the right. We can browse around the genome by clicking on
the left and right arrowm, or search for a specific gene with the
search text box.</p>
<p>This window also allows us to save a .png image of the canvas for
future reference if desired (i.e. Save Img button).</p>
</div>
<div class="section" id="scatter-plot">
<h2>Scatter plot<a class="headerlink" href="#scatter-plot" title="Permalink to this headline">¶</a></h2>
<p>We can also view a scatter plot of read counts of two selected
datasets. To achieve this we select two datasets (using &#8220;Control +
Clicck&#8221;) and then clicking on &#8220;View -&gt; Scatter Plot&#8221; in the menu bar
at the top of the TRANSIT window.</p>
<a class="reference internal image-reference" href="_images/transit_dataset_scatter_graph.png"><img alt="_images/transit_dataset_scatter_graph.png" class="align-center" src="_images/transit_dataset_scatter_graph.png" style="width: 600px;" /></a>
<p>A new window will pop-up, show a scatter plot of both of the selected
datasets. This window contains controls to zoom in and out (magnifying
glass), allowing us to focus in on a specific area. This is
particularly useful when large outliers may throw off the scale of the
scatter plot.</p>
</div>
<div class="section" id="essentiality-analysis-with-the-gumbel-method">
<h2>Essentiality analysis with the Gumbel method<a class="headerlink" href="#essentiality-analysis-with-the-gumbel-method" title="Permalink to this headline">¶</a></h2>
<p>Before comparing both conditions against each other, we may want to
determine which genes are essential in a specific condition to get an
idea of the genes which are required. To do this we can use the Gumbel
or the HMM methods, which determine essentiality within one condition.
First we chose the Gumbel method from the list of (Himar1) analysis methods
in the menu on top:</p>
<a class="reference internal image-reference" href="_images/transit_gumbel_options.png"><img alt="_images/transit_gumbel_options.png" class="align-center" src="_images/transit_gumbel_options.png" style="width: 600px;" /></a>
<p>For this particular case we leave the parameters at their default
settings as these work with a wide variety of datasets (See above for
an explanation of their function). We then click on the &#8220;Run Gumbel&#8221;
button and wait until the analysis finished running. The progress bar
will give us information about how much of the analysis is still left.
Once the program finishes, the results file is automatically created
(with the name chosen at run-time) and it is automatically added to
the Results File section at the bottom of TRANSIT. We can visualize
the results by selecting this file from the list, and clicking on the
&#8220;Display Table&#8221; button. This will open a new window with a table of
resuls:</p>
<p>From this window we can view results, and sort on a specific column
(described above) by clicking on a column header. In addition, the top
of this window contains a breakdown of the number of essential and
non-essential genes found by the Gumbel method. We can see that 675
genes are found to be essential by the Gumbel method (16%), roughly
matching expectations that 15% of the genomes is necessary for growth
in bacterial organisms. Clicking on the &#8220;Zbar&#8221; column we can sort the
data on the posterior probability of essentiality. If we sort in
descending order, we get those genes which are most likely to be
essential on the top. Among these are genes like GyrA (DNA gyrase A)
and RpoB (DNA-directed polymerase), which are both well-known
essential genes, and which are obtain a posterior probability of
essentiality of 1.0 (Essential).</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/transit_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial: Essentiality Analysis in a Single Condition</a><ul>
<li><a class="reference internal" href="#run-transit">Run TRANSIT</a></li>
<li><a class="reference internal" href="#adding-the-annotation-file">Adding the annotation file</a></li>
<li><a class="reference internal" href="#adding-the-control-datasets">Adding the control datasets</a></li>
<li><a class="reference internal" href="#visualizing-read-counts">Visualizing read counts</a></li>
<li><a class="reference internal" href="#scatter-plot">Scatter plot</a></li>
<li><a class="reference internal" href="#essentiality-analysis-with-the-gumbel-method">Essentiality analysis with the Gumbel method</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/transit_essentiality_tutorial.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">transit v2.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Michael A. DeJesus.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>